peft==0.6.1 implicitly requires transformers>=4.30.0, but transformers version is not specified

### System Info

Hi peft community,

We use `peft` in our project and yesterday tests started to fail in our CI with an error like:
```
/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/peft/peft_model.py:35: in <module>
    from transformers.pytorch_utils import id_tensor_storage
E   ImportError: cannot import name 'id_tensor_storage' from 'transformers.pytorch_utils' (/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/transformers/pytorch_utils.py)
```

Line `from transformers.pytorch_utils import id_tensor_storage` was added in peft==0.6.1 as a part of this PR by @younesbelkada: https://github.com/huggingface/peft/pull/1101/files#diff-95787fcc5356568929e383449ade2c1bac5da43deccab17318652b62ed171ae7R35

Function `id_tensor_storage` was added to `transformers` in version 4.30.0: https://github.com/huggingface/transformers/commit/d68d6665f94f86cbd9d5f512c9d2942528661cad

**So, it is required to have at least `transformers==4.30.0` to use `peft==0.6.1.`**.

**At the same time, `peft` does not set any version limits for `transformers` package: https://github.com/huggingface/peft/blob/main/setup.py#L46C10-L46C22**

In my project's `requirements.txt`, we have a pinned version `transformers==4.28.1`. Because `peft` does not specify `transformers` version, pip is fine to install `peft==0.6.1` and `transformers==4.28.1`, although it raises an ImportError in runtime later.

**Possible solution**: explicitly set `transformers>=0.6.1` in `setup.py`.

### Who can help?

@younesbelkada

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

```
conda create -n peft-reproduce python=3.10
conda activaate peft-reproduce
pip install transformers==4.28.1 peft==0.6.1
python
>>> import peft
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/anaconda3/envs/peft-reproduce/lib/python3.10/site-packages/peft/__init__.py", line 22, in <module>
    from .auto import (
  File "/opt/homebrew/anaconda3/envs/peft-reproduce/lib/python3.10/site-packages/peft/auto.py", line 31, in <module>
    from .mapping import MODEL_TYPE_TO_PEFT_MODEL_MAPPING
  File "/opt/homebrew/anaconda3/envs/peft-reproduce/lib/python3.10/site-packages/peft/mapping.py", line 23, in <module>
    from .peft_model import (
  File "/opt/homebrew/anaconda3/envs/peft-reproduce/lib/python3.10/site-packages/peft/peft_model.py", line 35, in <module>
    from transformers.pytorch_utils import id_tensor_storage
ImportError: cannot import name 'id_tensor_storage' from 'transformers.pytorch_utils' (/opt/homebrew/anaconda3/envs/peft-reproduce/lib/python3.10/site-packages/transformers/pytorch_utils.py)

### Expected behavior

`pip install` should fail if `transformers<4.30.0`

--- a/src/peft/import_utils.py
+++ b/src/peft/import_utils.py
@@ -14,6 +14,7 @@
 # limitations under the License.
 import importlib
 import importlib.metadata as importlib_metadata
+from functools import lru_cache
 
 import packaging.version
 
@@ -46,3 +47,20 @@ def is_auto_gptq_available():
 
 def is_optimum_available() -> bool:
     return importlib.util.find_spec("optimum") is not None
+
+
+@lru_cache()
+def is_torch_tpu_available(check_device=True):
+    "Checks if `torch_xla` is installed and potentially if a TPU is in the environment"
+    if importlib.util.find_spec("torch_xla") is not None:
+        if check_device:
+            # We need to check if `xla_device` can be found, will raise a RuntimeError if not
+            try:
+                import torch_xla.core.xla_model as xm
+
+                _ = xm.xla_device()
+                return True
+            except RuntimeError:
+                return False
+        return True
+    return False

--- a/src/peft/peft_model.py
+++ b/src/peft/peft_model.py
@@ -32,7 +32,6 @@ from safetensors.torch import save_file as safe_save_file
 from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, MSELoss
 from transformers import PreTrainedModel
 from transformers.modeling_outputs import QuestionAnsweringModelOutput, SequenceClassifierOutput, TokenClassifierOutput
-from transformers.pytorch_utils import id_tensor_storage
 from transformers.utils import PushToHubMixin
 
 from . import __version__
@@ -60,6 +59,7 @@ from .utils import (
     _set_adapter,
     _set_trainable,
     get_peft_model_state_dict,
+    id_tensor_storage,
     infer_device,
     load_peft_weights,
     set_peft_model_state_dict,

--- a/src/peft/utils/__init__.py
+++ b/src/peft/utils/__init__.py
@@ -45,6 +45,7 @@ from .other import (
     infer_device,
     get_auto_gptq_quant_linear,
     get_quantization_config,
+    id_tensor_storage,
 )
 from .hub_utils import hub_file_exists
 from .save_and_load import get_peft_model_state_dict, set_peft_model_state_dict, load_peft_weights

--- a/src/peft/utils/other.py
+++ b/src/peft/utils/other.py
@@ -15,14 +15,15 @@
 import copy
 import inspect
 import warnings
-from typing import Optional
+from typing import Optional, Tuple
 
 import accelerate
 import torch
 from accelerate.hooks import add_hook_to_module, remove_hook_from_module
 from accelerate.utils import is_npu_available, is_xpu_available
+from safetensors.torch import storage_ptr, storage_size
 
-from ..import_utils import is_auto_gptq_available
+from ..import_utils import is_auto_gptq_available, is_torch_tpu_available
 
 
 # Get current device name based on available devices
@@ -436,6 +437,31 @@ def get_auto_gptq_quant_linear(gptq_quantization_config):
         )
         return AutoGPTQQuantLinear
     return None
+
+
+def id_tensor_storage(tensor: torch.Tensor) -> Tuple[torch.device, int, int]:
+    """
+    Unique identifier to a tensor storage. Multiple different tensors can share the same underlying storage. For
+    example, "meta" tensors all share the same storage, and thus their identifier will all be equal. This identifier is
+    guaranteed to be unique and constant for this tensor's storage during its lifetime. Two tensor storages with
+    non-overlapping lifetimes may have the same id.
+
+    This method is the exact same copy of
+    https://github.com/huggingface/transformers/blob/main/src/transformers/pytorch_utils.py#L282C1-L300C58 but we added
+    it here manually to avoid import issue with old versions of transformers.
+    """
+    if tensor.device.type == "xla" and is_torch_tpu_available():
+        # NOTE: xla tensors dont have storage
+        # use some other unique id to distinguish.
+        # this is a XLA tensor, it must be created using torch_xla's
+        # device. So the following import is safe:
+        import torch_xla
+
+        unique_id = torch_xla._XLAC._xla_get_tensor_id(tensor)
+    else:
+        unique_id = storage_ptr(tensor)
+
+    return tensor.device, unique_id, storage_size(tensor)
 
 
 TRANSFORMERS_MODELS_TO_LORA_TARGET_MODULES_MAPPING = {

